include/AIToolbox/MDP/Algorithms/MCTS.hpp

```diff
@@ -268,13 +270,15 @@ namespace AIToolbox::MDP {
         return rew;
     }
 
-    template <typename M>
-    double MCTS<M>::rollout(size_t s, unsigned depth) {
+    template <typename M, typename Sel, typename ST, typename AT>
+    double MCTS<M, Sel, ST, AT>::rollout(ST s, unsigned depth) {
         double rew = 0.0, totalRew = 0.0, gamma = 1.0;
 
-        std::uniform_int_distribution<size_t> generator(0, A-1);
+        //std::uniform_int_distribution<size_t> generator(0, A-1);
         for ( ; depth < maxDepth_; ++depth ) {
-            std::tie( s, rew ) = model_.sampleSR( s, generator(rand_) );
+            auto allowedActions = model_.getAllowedActions(s);
```

Comment:
This is I think at the wrong level of abstraction. If you put this choice within UCB, then you can very easily define an UCB and UCBVariableAction classes which select the action for you depending on the model you are using. This has the advantage of not having to necessarily use a variable action model all the time (so we can also avoid touching MDP::Model), but instead defining a separate exploration method depending on our current constraints (whatever that may be). Like this you are forcing models to be variable actions, and you'll again have to modify MCTS again in the future if you want to handle different things in action selection.

Reply:
I partly agree with you. Yes, the selection of the actions is the job of ```UCB``` but the crafting of the set of actions ```UCB``` has to choose from is the job of ```MCTS``` (the number of children of a state node) with the help of the model. That being said, it's true that for now ```UCB``` also takes care of creating the action nodes, so why not this as well. I'll have a go at it.

Reply:
True, but rather than calling `getAllowedActions(s)` in MCTS (or whatever else we want to use), you simply call `auto action = ucb_.getAction(s, model)`. Then all the complexity and particular weird methods you want your model to have can be encapsulated within the UCB class you want to use. The idea is that you, as a user, are going to define both your model and your UCB, while using the library's MCTS implementation. This would be the visitor pattern if I'm not mistaken.

Reply:
Sure, but the problem is that ```rollout``` doesn't use the selection method (and should not). Maybe a ```getRandomAction``` in UCB for the rollout to use?

Reply:
That can definitely work. Alternatively one could specify a `PolicyInterface &` to use for rollouts (there should be a `RandomPolicy` around, maybe that could be the default..). I don't care too much for one or the other.

Reply:
I've done the former for now, but it seems your alternative is cleaner, I'll have a look at that before making another commit.